cognite/neat/core/extractors/rules_to_graphql.py

```diff
@@ -0,0 +1,151 @@
+    return invalid_names
+
+
+def repair_name(name: str, entity_type: str, fix_casing: bool = False) -> str:
+    """Repairs an entity name to conform to GraphQL naming convention"""
+
+    # Remove any non GraphQL compliant characters
+    repaired_string = re.sub(r"[^_a-zA-Z0-9/_]", "", name)
+
+    # Name must start with a letter or underscore
+    if repaired_string[0].isdigit():
+        repaired_string = f"_{repaired_string}"
+
+    if not fix_casing:
+        return repaired_string
+    # Property names must be camelCase
+    if entity_type == "property" and repaired_string[0].isupper():
+        return repaired_string[0].lower() + repaired_string[1:]
+    # Class names must be PascalCase
+    elif entity_type == "class" and repaired_string[0].islower():
+        return repaired_string[0].upper() + repaired_string[1:]
+    else:
+        return repaired_string
+
+
+def _remove_query_type(schema_string: str) -> str:
+    """Removes unnecessary Query type to conform to Cognite's GraphQL API"""
+    lines = schema_string.split("\n")
+
+    for _i, line in enumerate(lines):
+        if "}" in line:
+            break
+
+    return "\n".join(lines[_i + 2 :])
+
+
+def _get_graphql_schema_string(schema: GraphQLSchema) -> str:
+    return _remove_query_type(print_schema(schema))
+
+
+def rules2graphql_schema(
+    transformation_rules: TransformationRules, fix_names: bool = True, fix_casing: bool = False
+) -> str:
+    """Generates a GraphQL schema from an instance of TransformationRules
+
+    Parameters
+    ----------
+    transformation_rules : TransformationRules
+        TransformationRules object
+    fix_names : bool, optional
+        Whether to attempt to repair invalid entity names, by default True
+    fix_casing : bool, optional
+        Whether to attempt to fix casing of entity names, by default False
+
+    Returns
+    -------
+    str
+        GraphQL schema string
+    """
+    gql_type_definitions: dict = {}
+    invalid_names: set = get_invalid_names(transformation_rules.get_entity_names())
+
+    if invalid_names and not fix_names:
+        msg = "Entity names must only contain [_a-zA-Z0-9] characters and can start only with [_a-zA-Z]"
+        logging.error(f"{msg}, following entities {invalid_names} do not follow these rules!")
+        raise GraphQLError(f"{msg}, following entities {invalid_names} do not follow these rules!")
+    elif invalid_names and fix_names:
+        msg = "Entity names must only contain [_a-zA-Z0-9] characters and can start only with [_a-zA-Z]"
+        logging.warn(
+            f"{msg}, following entities {invalid_names} do not follow these rules! Attempting to repair names..."
+        )
+        warnings.warn(
+            f"{msg}, following entities {invalid_names} do not follow these rules! Attempting to repair names...",
+            stacklevel=2,
+        )
+
+    def _define_fields(property_definitions: pd.DataFrame) -> dict[str, GraphQLField]:
+        gql_field_definitions = {}
```

Comment:
Suggest moving this line closer to use, right before the for loop, line 138 before my golfing.

Reply:
@doctrino need more clarification for this one, I am not sure what you mean

Reply:
The variable `gql_field_definitions` is not used here, it is used furter down. Suggest moving it as close as possible to where you start populating it.

Reply:
That is more readable, as when I read the loop where it is used, I got had to scroll up here again to see which data structure it is.

Reply:
I see, but that would require logic to create this variable first time we create field, etc. , 
Anyways with new update of code I think it is clear, as at very beginning we check if property/field exist in this dict.